### PR TITLE
[5.7] Add Builder::insertSub()

### PR DESCRIPTION
**Pull request by miklcct**
This pull requests add the ability of inserting directly from a sub-query at SQL level, without loading data into memory.

Please refer to #2026 for its use.

### CODE REVIEW

**Comment by taylorotwell** [comment]
Is the syntax the same across all supported database drivers?

**Comment by miklcct** [comment]
They are the same in MySQL, SQL Server, PostgreSQL and SQLite.

Reference:
MySQL: https://dev.mysql.com/doc/refman/8.0/en/insert-select.html
SQL Server: https://docs.microsoft.com/en-us/sql/t-sql/statements/insert-transact-sql?view=sql-server-2017
PostgreSQL: https://www.postgresql.org/docs/9.1/sql-insert.html
SQLite: https://www.sqlite.org/lang_insert.html

File: src/Illuminate/Database/Query/Grammars/Grammar.php [hunk 1]
```diff
@@ -791,6 +791,23 @@ public function compileInsertGetId(Builder $query, $values, $sequence)
         return $this->compileInsert($query, $values);
     }
 
+    /**
+     * Compile an insert statement with subquery into SQL.
+     *
+     * @param  Builder  $query
+     * @param  array    $columns
+     * @param  string   $sql
```
Comment (driesvints) [inline src/Illuminate/Database/Query/Grammars/Grammar.php]: Use FQN here and proper spacing between parameters names

**Review by driesvints — CHANGES_REQUESTED** [review]

**Comment by taylorotwell** [comment]
Renamed to `insertUsing`.

File: src/Illuminate/Database/Query/Builder.php [hunk 2]
```diff
@@ -2558,6 +2558,23 @@ public function insertGetId(array $values, $sequence = null)
...
+     * @param  \Closure|\Illuminate\Database\Query\Builder|string  $query
+     * @return bool
+     */
+    public function insertSub(array $columns, $query)
+    {
+        [$sql, $bindings] = $this->createSub($query);
+
+        return $this->connection->insert(
+            $this->grammar->compileInsertSub($this, $columns, $sql),
+            $this->cleanBindings($bindings)
```
Comment (staudenmeir) [inline src/Illuminate/Database/Query/Builder.php]: Is there a case where we need `cleanBindings()`?

**Pull request merged**